File: src/lib/web-evidence.ts
/**
 * Web Evidence Search Module
 *
 * Searches PubMed and medical sources for real evidence to supplement
 * the RAG knowledge base. Used to ground appeal letters in verifiable,
 * current medical literature.
 *
 * All APIs used are free and require no API keys:
 * - PubMed E-utilities (NCBI)
 */

interface PubMedArticle {
    pmid: string;
    title: string;
    abstract: string;
    authors: string;
    journal: string;
    year: string;
    publicationTypes: string[];
}

interface WebEvidence {
    source: string;
    title: string;
    summary: string;
    citation: string;
    url: string;
    evidenceLevel: string;
}

/**
 * Search PubMed for relevant medical evidence.
 * Uses NCBI E-utilities API (free, no key required).
 */
async function searchPubMed(query: string, maxResults: number = 5): Promise<PubMedArticle[]> {
    try {
        // Step 1: Search for article IDs
        const searchUrl = `https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?db=pubmed&retmode=json&retmax=${maxResults}&sort=relevance&term=${encodeURIComponent(query)}`;
        const searchResponse = await fetch(searchUrl, { signal: AbortSignal.timeout(8000) });

        if (!searchResponse.ok) return [];

        const searchData = await searchResponse.json();
        const ids: string[] = searchData.esearchresult?.idlist || [];

        if (ids.length === 0) return [];

        // Step 2: Fetch article details
        const fetchUrl = `https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=pubmed&retmode=xml&id=${ids.join(',')}`;
        const fetchResponse = await fetch(fetchUrl, { signal: AbortSignal.timeout(8000) });

        if (!fetchResponse.ok) return [];

        const xml = await fetchResponse.text();
        return parsePubMedXml(xml);
    } catch (err) {
        console.error('PubMed search failed:', err);
        return [];
    }
}

/**
 * Parse PubMed XML response into structured articles.
 * Simple XML parsing without external dependencies.
 */
function parsePubMedXml(xml: string): PubMedArticle[] {
    const articles: PubMedArticle[] = [];
    const articleBlocks = xml.split('<PubmedArticle>').slice(1);

    for (const block of articleBlocks) {
        try {
            const pmid = extractXmlTag(block, 'PMID') || '';
            const title = extractXmlTag(block, 'ArticleTitle') || '';
            const abstractText = extractXmlTag(block, 'AbstractText') || '';
            const journal = extractXmlTag(block, 'Title') || '';
            const year = extractXmlTag(block, 'Year') || '';
            const publicationTypes = [...block.matchAll(/<PublicationType[^>]*>([^<]+)<\/PublicationType>/gi)]
                .map(match => cleanXmlText(match[1]));

            // Get first author
            const lastNameMatch = block.match(/<LastName>([^<]+)<\/LastName>/);
            const authors = lastNameMatch ? `${lastNameMatch[1]} et al.` : '';

            const isRetracted = publicationTypes.some(type => /retracted|retraction/i.test(type)) || /retracted publication/i.test(block);
            if (title && abstractText && !isRetracted) {
                articles.push({
                    pmid,
                    title: cleanXmlText(title),
                    abstract: cleanXmlText(abstractText).substring(0, 600),
                    authors,
                    journal: cleanXmlText(journal),
                    year,
                    publicationTypes,
                });
            }
        } catch {
            continue;
        }
    }

    return articles;
}

function extractXmlTag(xml: string, tag: string): string | null {
    // Handle tags that may have attributes
    const regex = new RegExp(`<${tag}[^>]*>([\\s\\S]*?)</${tag}>`, 'i');
    const match = xml.match(regex);
    return match ? match[1] : null;
}

function cleanXmlText(text: string): string {
    return text
        .replace(/<[^>]+>/g, '') // Remove nested XML tags
        .replace(/\s+/g, ' ')   // Normalize whitespace
        .trim();
}

/**
 * Build optimized PubMed search queries for a denied service.
 * Focuses on medical necessity evidence, clinical guidelines, and outcomes data.
 */
function buildSearchQueries(deniedService: string, denialReason: string, cptCodes: string, icd10Codes: string): string[] {
    const queries: string[] = [];
    const evidenceFilter = '(humans[MeSH Terms]) AND (systematic review[Publication Type] OR meta-analysis[Publication Type] OR practice guideline[Publication Type] OR randomized controlled trial[Publication Type] OR guideline[Publication Type])';

    // Primary: service + medical necessity + systematic review
    queries.push(`(${deniedService}) AND medical necessity AND ${evidenceFilter}`);

    // Evidence for the specific service
    queries.push(`(${deniedService}) AND clinical outcomes AND evidence-based AND ${evidenceFilter}`);

    // Address the denial reason specifically
    if (denialReason) {
        const shortReason = denialReason.substring(0, 100);
        queries.push(`(${deniedService}) AND (${shortReason}) AND coverage criteria AND humans[MeSH Terms]`);
    }
    if (cptCodes || icd10Codes) {
        queries.push(`(${deniedService}) AND (${cptCodes} ${icd10Codes}) AND medical policy AND humans[MeSH Terms]`);
    }

    return queries;
}

function evidenceLevel(publicationTypes: string[]): string {
    const joined = publicationTypes.join(' ').toLowerCase();
    if (/practice guideline|guideline/.test(joined)) return 'Guideline';
    if (/systematic review|meta-analysis/.test(joined)) return 'Systematic review/meta-analysis';
    if (/randomized controlled trial/.test(joined)) return 'Randomized controlled trial';
    if (/clinical trial/.test(joined)) return 'Clinical trial';
    return 'Peer-reviewed article';
}

const RELEVANCE_STOP_WORDS = new Set([
    'the', 'and', 'for', 'with', 'that', 'this', 'from', 'when', 'then', 'than', 'medical',
    'necessity', 'guidelines', 'guideline', 'coverage', 'criteria', 'clinical', 'outcomes',
    'evidence', 'based', 'requested', 'procedure', 'service', 'does', 'meet', 'not',
]);

function relevanceTokens(...values: string[]): string[] {
    return [...new Set(values
        .join(' ')
        .toLowerCase()
        .split(/[^a-z0-9]+/)
        .map(t => t.trim())
        .filter(t => t.length >= 4 && !RELEVANCE_STOP_WORDS.has(t) && !/^\d+$/.test(t)))];
}

function isRelevantArticle(article: PubMedArticle, deniedService: string, denialReason: string, icd10Codes: string): boolean {
    const serviceTokens = relevanceTokens(deniedService);
    const contextTokens = relevanceTokens(deniedService, denialReason, icd10Codes);
    const articleText = `${article.title} ${article.abstract}`.toLowerCase();
    const serviceMatches = serviceTokens.filter(token => articleText.includes(token)).length;
    const contextMatches = contextTokens.filter(token => articleText.includes(token)).length;

    // Require direct service relevance, or multiple broader context overlaps.
    return serviceMatches >= 1 || contextMatches >= 3;
}

/**
 * Main function: Search for web-based medical evidence to supplement RAG.
 * Returns formatted evidence that can be injected into the appeal prompt.
 */
export async function searchWebEvidence(
    deniedService: string,
    denialReason: string,
    cptCodes: string,
    icd10Codes: string,
): Promise<{ evidence: WebEvidence[]; formattedContext: string }> {
    const queries = buildSearchQueries(deniedService, denialReason, cptCodes, icd10Codes);

    // Search in parallel (but limit to avoid rate limits on PubMed)
    const allArticles: PubMedArticle[] = [];
    for (const query of queries) {
        const articles = await searchPubMed(query, 3);
        allArticles.push(...articles);

        // Small delay to be respectful to PubMed API
        if (queries.indexOf(query) < queries.length - 1) {
            await new Promise(r => setTimeout(r, 500));
        }
    }

    // Deduplicate by PMID
    const seen = new Set<string>();
    const uniqueArticles = allArticles.filter(a => {
        if (seen.has(a.pmid)) return false;
        seen.add(a.pmid);
        return true;
    }).filter(a => isRelevantArticle(a, deniedService, denialReason, icd10Codes));

    // Convert to WebEvidence format
    const evidence: WebEvidence[] = uniqueArticles.slice(0, 5).map(a => ({
        source: 'PubMed',
        title: a.title,
        summary: a.abstract,
        citation: `${a.authors} ${a.journal}. ${a.year}. PMID: ${a.pmid}`,
        url: `https://pubmed.ncbi.nlm.nih.gov/${a.pmid}/`,
        evidenceLevel: evidenceLevel(a.publicationTypes),
    }));

    // Format for prompt injection
    const formattedContext = evidence.length > 0
        ? evidence.map((e, i) =>
            `[PubMed ${i + 1}] ${e.title}\nEvidence Level: ${e.evidenceLevel}\nCitation: ${e.citation}\nURL: ${e.url}\n\nFindings: ${e.summary}`
        ).join('\n\n---\n\n')
        : '';

    return { evidence, formattedContext };
}

File: backend/app/services/web_evidence.py
import asyncio
import re
from dataclasses import dataclass

import httpx
import structlog

logger = structlog.get_logger()


@dataclass
class PubMedArticle:
    pmid: str
    title: str
    abstract: str
    authors: str
    journal: str
    year: str
    publication_types: list[str]


@dataclass
class WebEvidence:
    source: str
    title: str
    summary: str
    citation: str
    url: str
    evidence_level: str


def _extract_xml_tag(xml: str, tag: str) -> str | None:
    pattern = rf"<{tag}[^>]*>([\s\S]*?)</{tag}>"
    match = re.search(pattern, xml, re.IGNORECASE)
    return match.group(1) if match else None


def _clean_xml_text(text: str) -> str:
    text = re.sub(r"<[^>]+>", "", text)
    text = re.sub(r"\s+", " ", text)
    return text.strip()


def _parse_pubmed_xml(xml: str) -> list[PubMedArticle]:
    articles: list[PubMedArticle] = []
    blocks = xml.split("<PubmedArticle>")[1:]

    for block in blocks:
        try:
            pmid = _extract_xml_tag(block, "PMID") or ""
            title = _extract_xml_tag(block, "ArticleTitle") or ""
            abstract_text = _extract_xml_tag(block, "AbstractText") or ""
            journal = _extract_xml_tag(block, "Title") or ""
            year = _extract_xml_tag(block, "Year") or ""
            publication_types = [
                _clean_xml_text(m.group(1))
                for m in re.finditer(r"<PublicationType[^>]*>([^<]+)</PublicationType>", block, re.I)
            ]

            last_name_match = re.search(r"<LastName>([^<]+)</LastName>", block)
            authors = f"{last_name_match.group(1)} et al." if last_name_match else ""

            is_retracted = any(
                re.search(r"retracted|retraction", p, re.I) for p in publication_types
            ) or bool(re.search(r"retracted publication", block, re.I))

            if title and abstract_text and not is_retracted:
                articles.append(
                    PubMedArticle(
                        pmid=pmid,
                        title=_clean_xml_text(title),
                        abstract=_clean_xml_text(abstract_text)[:600],
                        authors=authors,
                        journal=_clean_xml_text(journal),
                        year=year,
                        publication_types=publication_types,
                    )
                )
        except Exception:
            continue

    return articles


async def _search_pubmed(
    client: httpx.AsyncClient, query: str, max_results: int = 5
) -> list[PubMedArticle]:
    try:
        search_resp = await client.get(
            "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi",
            params={"db": "pubmed", "retmode": "json", "retmax": max_results, "sort": "relevance", "term": query},
            timeout=8,
        )
        if not search_resp.is_success:
            return []

        ids = search_resp.json().get("esearchresult", {}).get("idlist", [])
        if not ids:
            return []

        fetch_resp = await client.get(
            "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi",
            params={"db": "pubmed", "retmode": "xml", "id": ",".join(ids)},
            timeout=8,
        )
        if not fetch_resp.is_success:
            return []

        return _parse_pubmed_xml(fetch_resp.text)
    except Exception as e:
        logger.warning("pubmed_search_failed", error=str(e))
        return []


def _build_search_queries(
    denied_service: str, denial_reason: str, cpt_codes: str, icd10_codes: str
) -> list[str]:
    evidence_filter = (
        "(humans[MeSH Terms]) AND (systematic review[Publication Type] OR "
        "meta-analysis[Publication Type] OR practice guideline[Publication Type] OR "
        "randomized controlled trial[Publication Type] OR guideline[Publication Type])"
    )
    queries = [
        f"({denied_service}) AND medical necessity AND {evidence_filter}",
        f"({denied_service}) AND clinical outcomes AND evidence-based AND {evidence_filter}",
    ]
    if denial_reason:
        short_reason = denial_reason[:100]
        queries.append(f"({denied_service}) AND ({short_reason}) AND coverage criteria AND humans[MeSH Terms]")
    if cpt_codes or icd10_codes:
        queries.append(f"({denied_service}) AND ({cpt_codes} {icd10_codes}) AND medical policy AND humans[MeSH Terms]")
    return queries


def _evidence_level(publication_types: list[str]) -> str:
    joined = " ".join(publication_types).lower()
    if "practice guideline" in joined or "guideline" in joined:
        return "Guideline"
    if "systematic review" in joined or "meta-analysis" in joined:
        return "Systematic review/meta-analysis"
    if "randomized controlled trial" in joined:
        return "Randomized controlled trial"
    if "clinical trial" in joined:
        return "Clinical trial"
    return "Peer-reviewed article"


RELEVANCE_STOP_WORDS = {
    "the", "and", "for", "with", "that", "this", "from", "when", "then", "than",
    "medical", "necessity", "guidelines", "guideline", "coverage", "criteria",
    "clinical", "outcomes", "evidence", "based", "requested", "procedure", "service",
    "does", "meet", "not",
}


def _relevance_tokens(*values: str) -> list[str]:
    tokens = [
        t.strip()
        for t in re.split(r"[^a-z0-9]+", " ".join(values).lower())
        if len(t.strip()) >= 4 and t.strip() not in RELEVANCE_STOP_WORDS and not t.strip().isdigit()
    ]
    return list(dict.fromkeys(tokens))


def _is_relevant_article(
    article: PubMedArticle,
    denied_service: str,
    denial_reason: str,
    icd10_codes: str,
) -> bool:
    service_tokens = _relevance_tokens(denied_service)
    context_tokens = _relevance_tokens(denied_service, denial_reason, icd10_codes)
    article_text = f"{article.title} {article.abstract}".lower()
    service_matches = sum(1 for token in service_tokens if token in article_text)
    context_matches = sum(1 for token in context_tokens if token in article_text)
    return service_matches >= 1 or context_matches >= 3


async def search_web_evidence(
    denied_service: str,
    denial_reason: str,
    cpt_codes: str,
    icd10_codes: str,
) -> dict:
    queries = _build_search_queries(denied_service, denial_reason, cpt_codes, icd10_codes)

    all_articles: list[PubMedArticle] = []
    async with httpx.AsyncClient() as client:
        for i, query in enumerate(queries):
            articles = await _search_pubmed(client, query, 3)
            all_articles.extend(articles)
            if i < len(queries) - 1:
                await asyncio.sleep(0.5)

    seen: set[str] = set()
    unique: list[PubMedArticle] = []
    for a in all_articles:
        if a.pmid not in seen:
            seen.add(a.pmid)
            if _is_relevant_article(a, denied_service, denial_reason, icd10_codes):
                unique.append(a)

    evidence: list[WebEvidence] = [
        WebEvidence(
            source="PubMed",
            title=a.title,
            summary=a.abstract,
            citation=f"{a.authors} {a.journal}. {a.year}. PMID: {a.pmid}",
            url=f"https://pubmed.ncbi.nlm.nih.gov/{a.pmid}/",
            evidence_level=_evidence_level(a.publication_types),
        )
        for a in unique[:5]
    ]

    formatted_context = ""
    if evidence:
        formatted_context = "\n\n---\n\n".join(
            f"[PubMed {i + 1}] {e.title}\nEvidence Level: {e.evidence_level}\nCitation: {e.citation}\nURL: {e.url}\n\nFindings: {e.summary}"
            for i, e in enumerate(evidence)
        )

    return {
        "evidence": [
            {
                "source": e.source,
                "title": e.title,
                "citation": e.citation,
                "url": e.url,
                "evidenceLevel": e.evidence_level,
            }
            for e in evidence
        ],
        "formattedContext": formatted_context,
    }
